### PR TITLE
[template] Augment dreambooth ft example with lora

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?
Augment dreambooth ft example with LoRA. 

@justinvyu per our discussion, I also removed the special model component placement logic for simplicity and new each worker only uses 1 GPU.

playground feel:

Screen Shot 2023-08-07 at 4 44 32 PM

Also added a release test and updated `doc/source/ray-air/examples/dreambooth_finetuning.rst`

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/flags.py [hunk 1]
```diff
@@ -17,7 +17,10 @@ def train_arguments():
         type=str,
         default=None,
         required=True,
-        help="Directory where trained models are saved.",
+        help="Directory where trained models or LoRA weights are saved.",
+    )
+    parser.add_argument(
+        "--use_lora", default=False, action="store_true", help="Use LoRA."
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/flags.py:23]: `help="Uses LoRA with default settings. The default settings are in foo.yaml"`

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/flags.py [hunk 2]
```diff
@@ -146,5 +149,10 @@ def run_model_flags():
             "Enable using Ray Data to use multiple GPU workers to perform inference."
         ),
     )
+    parser.add_argument(
+        "--lora_weights_dir",
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/flags.py:153]: If this is not passed in, can we assume it's the same as `model_dir`? I guess my question is that do we need this extra parameter? instead of re-using model_dir

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 3]
```diff
@@ -1,24 +1,42 @@
...
 import torch
 import torch.nn.functional as F
-from torch.nn.parallel import DistributedDataParallel
 from torch.nn.utils import clip_grad_norm_
 from transformers import CLIPTextModel
 
 from dataset import collate, get_train_dataset
 from flags import train_arguments
 
+LORA_RANK = 4
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:38]: can we create a json / yaml file for the default lora configs?

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 4]
```diff
@@ -48,7 +66,54 @@ def get_target(scheduler, noise, latents, timesteps):
...
+        unet_lora_attn_procs[name] = module
+        unet_lora_parameters.extend(module.parameters())
+
+    unet.set_attn_processor(unet_lora_attn_procs)
+
+    text_lora_parameters = LoraLoaderMixin._modify_text_encoder(
+        text_encoder, dtype=torch.float32, rank=LORA_RANK
+    )
+
+    return unet_lora_parameters, text_lora_parameters
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:113]: Naming confusion: Are these returns `unet_with_lora`, and `text_enc_with_lora`? Or just the parameters?

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 5]
```diff
@@ -48,7 +66,54 @@ def get_target(scheduler, noise, latents, timesteps):
     raise ValueError(f"Unknown prediction type {pred_type}")
 
 
-def load_models(config, cuda):
+def add_lora_layers(unet, text_encoder):
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:69]: Can we add a quick docstring, what the inputs are and what does this return?

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 6]
```diff
@@ -77,70 +140,78 @@ def load_models(config, cuda):
...
-    cuda = get_cuda_devices()
 
     # Load pre-trained models.
-    text_encoder, noise_scheduler, vae, unet = load_models(config, cuda)
+    (
+        text_encoder,
+        noise_scheduler,
+        vae,
+        unet,
+        unet_parameters,
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:175]: naming: Let's call these unet_trainable_params, and text_trainable_params

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 7]
```diff
@@ -77,70 +140,78 @@ def load_models(config, cuda):
...
+    for _ in range(num_train_epochs):
         if global_step >= config["max_train_steps"]:
             print(f"Stopping training after reaching {global_step} steps...")
             break
 
-        for step, batch in enumerate(
+        for _, batch in enumerate(
             train_dataset.iter_torch_batches(
-                batch_size=config["train_batch_size"], device=cuda[1]
+                # batch_size=config["train_batch_size"], device=cuda[1]
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:207]: remove the comment?

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 8]
```diff
@@ -77,70 +140,78 @@ def load_models(config, cuda):
...
+                batch_size=config["train_batch_size"],
+                device=train.torch.get_device(),
             )
         ):
             # Load batch on GPU 2 because VAE and text encoder are there.
-            batch = collate(batch, cuda[1], torch.bfloat16)
+            # batch = collate(batch, cuda[1], torch.bfloat16)
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:213]: remove the comment?

Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:213]: Does the comment above still apply?

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 9]
```diff
@@ -166,23 +237,22 @@ def train_fn(config):
             # Get the text embedding for conditioning
             encoder_hidden_states = text_encoder(batch["prompt_ids"])[0]
 
-            # Predict the noise residual. We need to move all data bits to GPU 1.
+            # Predit the noise residual.
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:240]: typo: Predict

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 10]
```diff
@@ -77,70 +140,78 @@ def load_models(config, cuda):
...
-    text_encoder = DistributedDataParallel(
-        text_encoder, device_ids=[cuda[1]], output_device=cuda[1]
-    )
-    unet = DistributedDataParallel(unet, device_ids=[cuda[0]], output_device=cuda[0])
+    text_encoder.train()
+    unet.train()
+
+    text_encoder = train.torch.prepare_model(text_encoder)
+    unet = train.torch.prepare_model(unet)
+    vae = vae.to(train.torch.get_device())
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:184]: As a user, I need clarification. So what does `train.torch.get_device()` return in this case? is it cuda:1? Why did we not have to do it for `unet`? Also for data why do we do `to(cuda)`?

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 11]
```diff
@@ -193,20 +263,55 @@ def train_fn(config):
...
     # Create pipeline using the trained modules and save it.
-    if train.get_context().get_world_rank() == 0:
-        pipeline = DiffusionPipeline.from_pretrained(
-            config["model_dir"],
-            text_encoder=text_encoder.module,
-            unet=unet.module,
-        )
-        pipeline.save_pretrained(config["output_dir"])
+    if session.get_world_rank() == 0:
+        if not config["use_lora"]:
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:274]: nit: I always use positive conditions whereever I can:

`if config["use_lora"] else`

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 12]
```diff
@@ -193,20 +263,55 @@ def train_fn(config):
...
+                text_encoder=text_encoder.module,
+                unet=unet.module,
+            )
+            pipeline.save_pretrained(config["output_dir"])
+        else:
+            save_lora_weights(unet.module, text_encoder.module, config["output_dir"])
+
+
+def unet_attn_processors_state_dict(unet) -> Dict[str, torch.tensor]:
+    r"""
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:286]: why is there an `r`?

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 13]
```diff
@@ -193,20 +263,55 @@ def train_fn(config):
...
+    Returns:
+        a state dict containing just the attention processor parameters.
+    """
+    attn_processors = unet.attn_processors
+
+    attn_processors_state_dict = {}
+
+    for attn_processor_key, attn_processor in attn_processors.items():
+        for parameter_key, parameter in attn_processor.state_dict().items():
+            attn_processors_state_dict[
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:296]: nit: ugly, let's do 
```
param_name = f"{attn_processor_key}.{parameter_key}"
attn_processors_state_dict[param_name] = parameter
```

File: doc/source/templates/05_dreambooth_finetuning/playground.ipynb [hunk 14]
```diff
@@ -38,13 +41,28 @@
...
-    "    TUNED_MODEL_PATH, torch_dtype=torch.float16\n",
-    ")\n",
-    "pipeline.to(\"cuda\")\n"
+    "from dreambooth.generate_utils import load_lora_weights, get_pipeline\n",
+    "\n",
+    "pipeline = get_pipeline(TUNED_MODEL_PATH)  # default to full finetuning\n",
+    "pipeline.to(\"cuda\")\n",
+    "\n",
+    "\n",
+    "def on_full_ft():\n",
```
Comment (kouroshHakha) [inline doc/source/templates/05_dreambooth_finetuning/playground.ipynb:55]: can we avoid using global? Make these functions take a pipeline variable.

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py (hunk 4 above)
Comment (xwjiang2010) [reply doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:113]: These only *returns* parameters and the model is modified *in place*.

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py (hunk 6 above)
Comment (xwjiang2010) [reply doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:175]: good suggestion, thanks!

File: doc/source/templates/05_dreambooth_finetuning/README.md [hunk 15]
```diff
@@ -31,11 +31,27 @@ Here are a few modifications to the `dreambooth_run.sh` script that you may want
...
+3. The `$NUM_WORKERS` variable sets the number of data-parallel workers used during fine-tuning. The default is 2 workers (2 workers, each using 1 GPU), and you should increase this number if you add more GPU worker nodes to the cluster.
 4. Setting `--num_epochs` and `--max_train_steps` determines the number of fine-tuning steps to take.
     - Depending on the batch size and number of data-parallel workers, one epoch will run for a certain number of steps. The run will terminate when one of these values (epoch vs. total number of steps) is reached.
 5. `generate.py` is used to generate stable diffusion images after loading the model from a checkpoint. You should modify the prompt at the end to be something more interesting, rather than just a photo of your subject.
 6. If you want to launch another fine-tuning run, you may want to run *only* the `python train.py ...` command. Running the bash script will start from the beginning (generating another regularization dataset).
+7. Use the following command for LoRA fine-tuning.
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply doc/source/templates/05_dreambooth_finetuning/README.md:39]: updated with --use_lora and --skip_image_download

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/flags.py (hunk 1 above)
Comment (xwjiang2010) [reply doc/source/templates/05_dreambooth_finetuning/dreambooth/flags.py:23]: going to leave this as a TODO if ok with you. Right now, probably only rank and learning rate. If we have more sophisticated setups, I will move to a dedicated yaml file.

**Review by kouroshHakha — APPROVED** [review]

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py (hunk 10 above)
Comment (kouroshHakha) [reply doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:200]: Can we leave comments for this explanation?

**Comment by xwjiang2010** [comment]
release test @ https://buildkite.com/ray-project/release-tests-pr/builds/48129

**Review by kouroshHakha — APPROVED** [review]

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py [hunk 16]
```diff
@@ -166,23 +252,22 @@ def train_fn(config):
...
 
-            # Predict the noise residual. We need to move all data bits to GPU 1.
+            # Predict the noise residual.
             model_pred = unet(
-                noisy_latents.to(cuda[0]),
-                timesteps.to(cuda[0]),
-                encoder_hidden_states.to(cuda[0]),
+                noisy_latents.cuda(),
+                timesteps.cuda(),
+                encoder_hidden_states.cuda(),
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:259]: Example on 4 GPU machine:

```python
from ray import air, train

import os
import torch
from ray.train.torch import TorchTrainer, get_device


def train_fn(config):
    print("cuda visible devices:", os.environ.get("CUDA_VISIBLE_DEVICES"))
    print("torch cuda current device:", torch.cuda.current_device())
    print("train.get_device:", get_device())
    print(torch.ones(1).cuda())

trainer = TorchTrainer(train_fn, scaling_config=air.ScalingConfig(num_workers=2, use_gpu=True))
trainer.fit()
```

```
(RayTrainWorker pid=17008) cuda visible devices: 0,1
(RayTrainWorker pid=17008) torch cuda current device: 0
(RayTrainWorker pid=17008) train.get_device: cuda:1
(RayTrainWorker pid=17007) cuda visible devices: 0,1
(RayTrainWorker pid=17007) torch cuda current device: 0
(RayTrainWorker pid=17007) train.get_device: cuda:0
(RayTrainWorker pid=17008) tensor([1.], device='cuda:0')
(RayTrainWorker pid=17007) tensor([1.], device='cuda:0')
```

**Review by justinvyu — CHANGES_REQUESTED** [review]
Some small fixes. Also, forgot to merge https://github.com/ray-project/ray/pull/38180, could you also rebase once that one is in? TY

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py:259]: yeah interesting. I think you are right. `train.torch.get_device` is the right way to go. However, I don't see different device exception with this bug. Let me try out with ws that I have.

**Comment by xwjiang2010** [comment]
https://buildkite.com/ray-project/release-tests-pr/builds/48188

**Review by justinvyu — APPROVED** [review]
Thanks!

**Comment by xwjiang2010** [comment]
Ok, finally Cuong helped me find a way to trigger release test without building the wheel and here it is: https://buildkite.com/ray-project/release-tests-pr/builds/48245#0189d835-2d2e-46ef-9bc3-205eba353c24

File: doc/source/ray-air/examples/dreambooth_finetuning.rst [hunk 17]
```diff
@@ -16,8 +16,8 @@ See the HuggingFace tutorial for useful explanations and suggestions on hyperpar
 
 **Compute requirements:**
 
-* Because of the large model sizes, you'll need a machine with at least 2 A10G GPUs.
-* Each training worker uses 2 GPUs, so you'll need ``N * 2`` total GPUs in your Ray cluster, if training with ``N`` distributed training workers. The example can leverage data-parallel training with more GPUs to speed up training time.
+* Because of the large model sizes, you'll need a machine with at least 1 A10G GPU.
```
Comment (matthewdeng) [inline doc/source/ray-air/examples/dreambooth_finetuning.rst:19]: Quick question, why does this not need 2 GPUs anymore? If this is only applicable when LoRA is enabled I would be explicit about this.

**Review by matthewdeng — APPROVED** [review]
Stamp

File: doc/source/ray-air/examples/dreambooth_finetuning.rst (hunk 17 above)
Comment (xwjiang2010) [reply doc/source/ray-air/examples/dreambooth_finetuning.rst:19]: No applicable to non-lora case as well. We decided it this way, so the example can directly use `prepare_model` and not a lot of manual stuff. Less noise.

**Comment by xwjiang2010** [comment]
release test: https://buildkite.com/ray-project/release-tests-pr/builds/48865#0189f55a-de38-4ca6-9fb7-3b497c9ee84f

File: doc/source/ray-air/examples/dreambooth_finetuning.rst [hunk 18]
```diff
@@ -147,30 +147,30 @@ We ran training using 2, and 4 workers/GPUs to compare throughput.
 
 Setup:
 
-* One g5.12xlarge nodes with 4 A10G GPUs each
+* One g5.12xlarge nodes with 4 A10G GPUs
 * Model as configured below
 * Data from this example
 * 200 regularization images
 * Training for 4 epochs (800 steps)
+* 3 runs per configuration
```
Comment (justinvyu) [inline doc/source/ray-air/examples/dreambooth_finetuning.rst:155]: Do we have std deviation / error bar for the plot?

**Review by justinvyu — COMMENTED** [review]
Thanks! Just some questions.

File: doc/source/templates/05_dreambooth_finetuning/dreambooth/images/dreambooth_training.png
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply doc/source/templates/05_dreambooth_finetuning/dreambooth/images/dreambooth_training.png:1]: the numbers were generated using gce instance. I just switched back to aws, the number for a 2 worker 2 GPU setup is only 289 compared with the original 364.

File: doc/source/ray-air/examples/dreambooth_finetuning.rst (hunk 18 above)
Comment (xwjiang2010) [reply doc/source/ray-air/examples/dreambooth_finetuning.rst:155]: not much value to the visual. std dev is super small and almost not visible.

**Comment by xwjiang2010** [comment]
CI failures are irrelevant. Merging.
